src/permissions.cxx

```diff
@@ -0,0 +1,161 @@
+                                              Permissions::Nowhere};
+  std::string best_candidate_name = "";
+  int max_len = 0;
+  for (const auto& [varname, rights] : variable_permissions) {
+    if (varname == variable) {
+      return {varname, rights};
+    }
+    if (varname.size() > max_len and variable.find(varname + ":") == 0) {
+      max_len = varname.size();
+      best_candidate = rights;
+      best_candidate_name = varname;
+    }
+  }
+  return {best_candidate_name, best_candidate};
+}
+
+std::pair<bool, std::string> Permissions::canAccess(const std::string& variable,
+                                                    PermissionTypes permission,
+                                                    Regions region) const {
+  auto [match_name, match_rights] = bestMatchRights(variable);
+  if ((match_rights[permission] & region) == region) {
+    return {true, match_name};
+  } else {
+    return {false, ""};
+  }
+}
+
+std::pair<Permissions::PermissionTypes, std::string>
+Permissions::getHighestPermission(const std::string& variable,
+                                  Permissions::Regions region) const {
+  if (region == Nowhere)
+    return {None, ""};
+  auto [varname, rights] = bestMatchRights(variable);
+  int i = ReadIfSet;
+  while (i < PERMISSION_TYPES_END and (rights[i] & region) == region) {
+    i++;
+  }
+  return {static_cast<PermissionTypes>(i - 1), varname};
+}
+
+std::map<std::string, Permissions::Regions>
+Permissions::getVariablesWithPermission(PermissionTypes permission,
+                                        bool highestOnly) const {
+  std::map<std::string, Permissions::Regions> result;
+  if (highestOnly and permission < PERMISSION_TYPES_END - 1) {
+    for (const auto& [varname, rights] : variable_permissions) {
+      auto regions = rights[permission];
+      auto perm_in_regions =
+          static_cast<Regions>(rights[permission] & ~rights[permission + 1]);
+      if (perm_in_regions != Nowhere)
+        result.emplace(varname, perm_in_regions);
+    }
+  } else {
+    for (const auto& [varname, rights] : variable_permissions) {
+      auto regions = rights[permission];
+      if (regions != Nowhere)
+        result.emplace(varname, regions);
+    }
+  }
+  return result;
+}
+
+Permissions::AccessRights Permissions::applyLowerPermissions(const AccessRights& rights) {
+  AccessRights result(rights);
+  for (int i = ReadIfSet; i < PERMISSION_TYPES_END; i++) {
+    result[i] = rights[i];
+    // Higher permissions imply lower permissions
+    for (int j = ReadIfSet; j < i; j++) {
+      result[j] = static_cast<Regions>(result[j] | rights[i]);
+    }
+  }
+  return result;
+}
+
+std::string Permissions::regionNames(const Regions regions) {
+  std::string result;
+  for (auto & [region, name] : fundamental_regions) {
+    if ((regions & region) == region) {
+      if (result.size() > 0) result += ", ";
+      result += name;
```

Comment:
This looks like `fmt::join`?

Reply:
Yes, I guess it is. I wasn't aware of that function.